File: libs/LLMKit/Sources/LLMKit/Providers/Anthropic.swift
import Foundation

public struct Anthropic: Sendable {
    public let baseURL: String

    public init(baseURL: String) {
        self.baseURL = baseURL
    }

    public static func api(baseURL: String) -> API {
        .anthropic(Anthropic(baseURL: baseURL))
    }

    public func listModels(token: String) async throws -> [ModelDesc] {
        struct AnthropicModelList: Decodable {
            let models: [ModelDesc]
        }

        let (response, data) = try await HTTPClient.request(
            method: "GET",
            url: try endpoint("/v1/models"),
            headers: baseHeaders(token: token)
        )

        guard (200..<300).contains(response.statusCode) else {
            throw LLMError.serverError(response.statusCode)
        }

        do {
            return try JSONDecoder().decode(AnthropicModelList.self, from: data).models
        } catch {
            throw LLMError.parsingError(String(describing: error))
        }
    }

    public func getModel(token: String, model: String) async throws -> ModelDesc {
        let (response, data) = try await HTTPClient.request(
            method: "GET",
            url: try endpoint("/v1/models/\(model)"),
            headers: baseHeaders(token: token)
        )

        guard (200..<300).contains(response.statusCode) else {
            throw LLMError.serverError(response.statusCode)
        }

        do {
            return try JSONDecoder().decode(ModelDesc.self, from: data)
        } catch {
            throw LLMError.parsingError(String(describing: error))
        }
    }

    public func getCompletion(token: String, request: CompletionRequest) async throws -> Completion {
        let upstream = try await postMessages(
            token: token,
            body: try messagesRequestBody(
                model: request.model,
                maxTokens: request.maxTokens ?? 8192,
                messages: request.messages,
                stream: nil
            )
        )

        let choices = upstream.content.enumerated().map { index, content in
            CompletionChoice(
                message: nil,
                text: content.text,
                index: UInt16(index),
                delta: nil,
                logprobs: nil,
                finishReason: upstream.stopReason
            )
        }

        let prompt = upstream.usage.inputTokens
        let completion = upstream.usage.outputTokens
        return Completion(
            id: upstream.id,
            created: 0,
            model: upstream.model,
            choices: choices,
            usage: Usage(
                promptTokens: prompt,
                completionTokens: completion,
                totalTokens: prompt + completion
            )
        )
    }

    public func streamCompletion(token: String, request: CompletionRequest) -> AsyncThrowingStream<StreamResponseChunk, Error> {
        let body: Data
        do {
            body = try messagesRequestBody(
                model: request.model,
                maxTokens: request.maxTokens ?? 1024,
                messages: request.messages,
                stream: true
            )
        } catch {
            return failedStream(error)
        }

        return streamMessageChunks(token: token, requestBody: body) { chunk in
            guard let text = chunk.textDelta else { return nil }
            return StreamResponseChunk(
                id: "cmpl-\(UUID().uuidString.lowercased())",
                object: "chat.completion.chunk",
                created: Int(Date().timeIntervalSince1970),
                model: request.model,
                systemFingerprint: nil,
                choices: [
                    CompletionChoice(
                        message: nil,
                        text: text,
                        index: UInt16(chunk.index ?? 0),
                        delta: nil,
                        logprobs: nil,
                        finishReason: nil
                    )
                ]
            )
        }
    }

    public func getResponse(token: String, request: ResponseRequest) async throws -> Response {
        let upstream = try await postMessages(
            token: token,
            body: try messagesRequestBody(
                model: request.model,
                maxTokens: request.maxOutputTokens ?? 8192,
                messages: request.input,
                stream: nil
            )
        )

        let content = upstream.content.map { ResponseContent(type: $0.type, text: $0.text) }
        let prompt = upstream.usage.inputTokens
        let completion = upstream.usage.outputTokens

        return Response(
            id: upstream.id,
            model: upstream.model,
            output: [ResponseOutput(type: "message", role: .assistant, content: content)],
            usage: Usage(
                promptTokens: prompt,
                completionTokens: completion,
                totalTokens: prompt + completion
            )
        )
    }

    public func streamResponse(token: String, request: ResponseRequest) -> AsyncThrowingStream<ResponseStreamEvent, Error> {
        let body: Data
        do {
            body = try messagesRequestBody(
                model: request.model,
                maxTokens: request.maxOutputTokens ?? 1024,
                messages: request.input,
                stream: true
            )
        } catch {
            return failedStream(error)
        }

        return streamMessageChunks(token: token, requestBody: body) { chunk in
            guard let text = chunk.textDelta else {
                if chunk.type == "message_stop" {
                    return ResponseStreamEvent(
                        type: "response.completed",
                        responseID: nil,
                        outputIndex: nil,
                        delta: nil,
                        text: nil
                    )
                }
                return nil
            }

            return ResponseStreamEvent(
                type: "response.output_text.delta",
                responseID: nil,
                outputIndex: chunk.index,
                delta: text,
                text: nil
            )
        }
    }

    private func postMessages(token: String, body: Data) async throws -> AnthropicMessageResponse {
        var headers = baseHeaders(token: token)
        headers["Content-Type"] = "application/json"

        let (response, data) = try await HTTPClient.request(
            method: "POST",
            url: try endpoint("/v1/messages"),
            headers: headers,
            body: body
        )

        guard (200..<300).contains(response.statusCode) else {
            throw LLMError.serverError(response.statusCode)
        }

        do {
            return try JSONDecoder().decode(AnthropicMessageResponse.self, from: data)
        } catch {
            throw LLMError.parsingError(String(describing: error))
        }
    }

    private func streamMessageChunks<T: Sendable>(
        token: String,
        requestBody: Data,
        map: @escaping @Sendable (AnthropicStreamChunk) -> T?
    ) -> AsyncThrowingStream<T, Error> {
        AsyncThrowingStream { continuation in
            Task {
                do {
                    var headers = baseHeaders(token: token)
                    headers["Content-Type"] = "application/json"

                    let lineStream = HTTPClient.streamLines(
                        method: "POST",
                        url: try endpoint("/v1/messages"),
                        headers: headers,
                        body: requestBody
                    )

                    for try await line in lineStream {
                        let trimmed = line.trimmingCharacters(in: .whitespacesAndNewlines)
                        guard trimmed.hasPrefix("data: ") else { continue }

                        let payload = String(trimmed.dropFirst(6))
                        if payload == "[DONE]" {
                            continuation.finish()
                            return
                        }

                        guard let data = payload.data(using: .utf8) else { continue }
                        guard let chunk = try? JSONDecoder().decode(AnthropicStreamChunk.self, from: data) else { continue }

                        if let mapped = map(chunk) {
                            continuation.yield(mapped)
                            if chunk.type == "message_stop" {
                                continuation.finish()
                                return
                            }
                        }
                    }
                    continuation.finish()
                } catch {
                    continuation.finish(throwing: error)
                }
            }
        }
    }

    private func messagesRequestBody(model: String, maxTokens: UInt32, messages: [Message], stream: Bool?) throws -> Data {
        try JSONEncoder().encode(
            AnthropicMessagesRequest(
                model: model,
                maxTokens: maxTokens,
                messages: messages,
                system: nil,
                stream: stream
            )
        )
    }

    private func failedStream<T>(_ error: Error) -> AsyncThrowingStream<T, Error> {
        AsyncThrowingStream { continuation in
            continuation.finish(throwing: error)
        }
    }

    private func baseHeaders(token: String) -> [String: String] {
        [
            "x-api-key": token,
            "anthropic-version": "2023-06-01"
        ]
    }

    private func endpoint(_ path: String) throws -> URL {
        guard let url = URL(string: "\(baseURL.trimmingCharacters(in: CharacterSet(charactersIn: "/")))\(path)") else {
            throw LLMError.invalidRequest("invalid URL")
        }
        return url
    }
}

private struct AnthropicMessagesRequest: Encodable {
    let model: String
    let maxTokens: UInt32
    let messages: [Message]
    let system: String?
    let stream: Bool?

    enum CodingKeys: String, CodingKey {
        case model
        case maxTokens = "max_tokens"
        case messages
        case system
        case stream
    }
}

private struct AnthropicTextContent: Decodable {
    let type: String
    let text: String
}

private struct AnthropicUsage: Decodable {
    let inputTokens: UInt32
    let outputTokens: UInt32

    enum CodingKeys: String, CodingKey {
        case inputTokens = "input_tokens"
        case outputTokens = "output_tokens"
    }
}

private struct AnthropicMessageResponse: Decodable {
    let id: String
    let model: String
    let content: [AnthropicTextContent]
    let stopReason: String?
    let usage: AnthropicUsage

    enum CodingKeys: String, CodingKey {
        case id
        case model
        case content
        case stopReason = "stop_reason"
        case usage
    }
}

private struct AnthropicStreamChunk: Decodable, Sendable {
    let type: String?
    let index: UInt32?
    let delta: AnthropicDelta?
    let contentBlock: AnthropicDelta?
    let text: String?

    enum CodingKeys: String, CodingKey {
        case type
        case index
        case delta
        case contentBlock = "content_block"
        case text
    }

    var textDelta: String? {
        text ?? delta?.text ?? contentBlock?.text
    }
}

private struct AnthropicDelta: Decodable, Sendable {
    let text: String?
}

File: libs/llm/src/anthropic.rs
use std::pin::Pin;

use async_stream::stream;
use bytes::Bytes;
use futures::Stream;
use http_body_util::{BodyExt, Full};
use hyper::header::CONTENT_TYPE;
use hyper::{Method, Request};
use hyper_rustls::ConfigBuilderExt;
use hyper_util::client::legacy::Client;
use hyper_util::rt::TokioExecutor;
use serde::{Deserialize, Serialize};
use uuid;

use crate::completion_request::CompletionRequest;
use crate::{API, Completion, CompletionChoice, Error, Message, ModelDesc, StreamResponseChunk};

#[derive(Debug, Clone)]
pub struct Anthropic {
    base_url: String,
}

#[derive(Deserialize)]
struct AnthropicTextContent {
    r#type: String,
    text: String,
}

#[derive(Deserialize)]
struct AnthropicCompletionResponse {
    id: String,
    r#type: String,
    role: String,
    model: String,
    content: Vec<AnthropicTextContent>,
    stop_reason: Option<String>,
    stop_sequence: Option<String>,
    usage: AnthropicUsage,
}

#[derive(Deserialize)]
struct AnthropicUsage {
    input_tokens: u32,
    output_tokens: u32,
}

#[derive(Serialize)]
struct AnthropicCompletionRequest<'a> {
    model: &'a str,
    max_tokens: u32,
    messages: &'a [Message],
    #[serde(skip_serializing_if = "Option::is_none")]
    system: Option<&'a str>,
    #[serde(skip_serializing_if = "Option::is_none")]
    stream: Option<bool>,
}

impl Anthropic {
    pub fn new(base_url: &str) -> API {
        API::Anthropic(Anthropic {
            base_url: base_url.to_string(),
        })
    }

    pub async fn list_models(&self, token: &str) -> Result<Vec<ModelDesc>, Error> {
        #[derive(Debug, Deserialize)]
        struct AnthropicModelList {
            models: Vec<ModelDesc>,
        }

        let url = format!("{}/v1/models", self.base_url.trim_end_matches('/'));

        let tls = rustls::ClientConfig::builder()
            .with_native_roots()
            .map_err(|_| Error::Unknown)?
            .with_no_client_auth();

        let https = hyper_rustls::HttpsConnectorBuilder::new()
            .with_tls_config(tls)
            .https_or_http()
            .enable_http1()
            .build();

        let req = Request::builder()
            .method(Method::GET)
            .uri(url)
            .header("x-api-key", token)
            .header("anthropic-version", "2023-06-01")
            .body(Full::new(Bytes::new()))
            .map_err(|_| Error::Unknown)?;

        let client: Client<_, Full<Bytes>> = Client::builder(TokioExecutor::new()).build(https);
        let res = client.request(req).await.map_err(|_| Error::Unknown)?;
        let status = res.status();

        let body_bytes = res
            .into_body()
            .collect()
            .await
            .map_err(|_| Error::Unknown)?
            .to_bytes();

        if !status.is_success() {
            return Err(Error::Unknown);
        }

        let model_list: AnthropicModelList =
            serde_json::from_slice(&body_bytes).map_err(|_| Error::Unknown)?;

        Ok(model_list.models)
    }

    pub async fn get_model(&self, token: &str, model: &str) -> Result<ModelDesc, Error> {
        let url = format!(
            "{}/v1/models/{}",
            self.base_url.trim_end_matches('/'),
            model
        );

        let tls = rustls::ClientConfig::builder()
            .with_native_roots()
            .map_err(|_| Error::Unknown)?
            .with_no_client_auth();

        let https = hyper_rustls::HttpsConnectorBuilder::new()
            .with_tls_config(tls)
            .https_or_http()
            .enable_http1()
            .build();

        let req = Request::builder()
            .method(Method::GET)
            .uri(url)
            .header("x-api-key", token)
            .header("anthropic-version", "2023-06-01")
            .body(Full::new(Bytes::new()))
            .map_err(|_| Error::Unknown)?;

        let client: Client<_, Full<Bytes>> = Client::builder(TokioExecutor::new()).build(https);
        let res = client.request(req).await.map_err(|_| Error::Unknown)?;
        let status = res.status();

        let body_bytes = res
            .into_body()
            .collect()
            .await
            .map_err(|_| Error::Unknown)?
            .to_bytes();

        if !status.is_success() {
            return Err(Error::Unknown);
        }

        let model_desc: ModelDesc =
            serde_json::from_slice(&body_bytes).map_err(|_| Error::Unknown)?;

        Ok(model_desc)
    }

    pub async fn get_completion(
        &self,
        token: &str,
        request: CompletionRequest,
    ) -> Result<Completion, Error> {
        let url = format!("{}/v1/messages", self.base_url.trim_end_matches('/'));

        let req_body = AnthropicCompletionRequest {
            model: &request.model,
            max_tokens: request.max_tokens.unwrap_or(8192),
            messages: &request.messages,
            system: None,
            stream: None,
        };

        let json = serde_json::to_string(&req_body).map_err(|_| Error::Unknown)?;

        let req = Request::builder()
            .method(Method::POST)
            .uri(url)
            .header("x-api-key", token)
            .header("anthropic-version", "2023-06-01")
            .header(CONTENT_TYPE, "application/json")
            .body(Full::new(Bytes::from(json)))
            .map_err(|_| Error::Unknown)?;

        let tls = rustls::ClientConfig::builder()
            .with_native_roots()
            .map_err(|_| Error::Unknown)?
            .with_no_client_auth();

        let https = hyper_rustls::HttpsConnectorBuilder::new()
            .with_tls_config(tls)
            .https_or_http()
            .enable_http1()
            .build();

        let client: Client<_, Full<Bytes>> = Client::builder(TokioExecutor::new()).build(https);
        let res = client.request(req).await.map_err(|_| Error::Unknown)?;
        let status = res.status();

        let body_bytes = res
            .into_body()
            .collect()
            .await
            .map_err(|_| Error::Unknown)?
            .to_bytes();

        if !status.is_success() {
            return Err(Error::Unknown);
        }

        let anthropic: AnthropicCompletionResponse =
            serde_json::from_slice(&body_bytes).map_err(|_| Error::Unknown)?;

        let choices = anthropic
            .content
            .iter()
            .enumerate()
            .map(|(i, c)| CompletionChoice {
                message: None,
                text: Some(c.text.clone()),
                index: i as u16,
                delta: None,
                logprobs: None,
                finish_reason: anthropic.stop_reason.clone(),
            })
            .collect();

        Ok(Completion {
            id: anthropic.id,
            created: 0,
            model: anthropic.model,
            choices,
            usage: crate::Usage {
                prompt_tokens: anthropic.usage.input_tokens,
                completion_tokens: anthropic.usage.output_tokens,
                total_tokens: anthropic.usage.input_tokens + anthropic.usage.output_tokens,
            },
        })
    }

    pub fn stream_completion(
        &self,
        token: &str,
        request: CompletionRequest,
    ) -> Pin<Box<dyn Stream<Item = Result<StreamResponseChunk, Error>> + Send + 'static>> {
        #[derive(Deserialize)]
        struct StreamChunk {
            index: Option<u32>,
            delta: Option<AnthropicTextContent>,
            content_block: Option<AnthropicTextContent>,
            text: Option<String>,
        }

        let base_url = self.base_url.clone();
        let token = token.to_owned();

        let s = stream! {
            let url = format!("{}/v1/messages", base_url.trim_end_matches('/'));

            let req_body = AnthropicCompletionRequest {
                model: &request.model,
                max_tokens: request.max_tokens.unwrap_or(1024),
                messages: &request.messages,
                system: None,
                stream: Some(true),
            };

            let json = serde_json::to_string(&req_body)
                .map_err(|_| Error::Unknown)?;

            let req = Request::builder()
                .method(Method::POST)
                .uri(url)
                .header("x-api-key", token)
                .header("anthropic-version", "2023-06-01")
                .header(CONTENT_TYPE, "application/json")
                .body(Full::new(Bytes::from(json)))
                .map_err(|_| Error::Unknown)?;

            let tls = rustls::ClientConfig::builder()
                .with_native_roots()
                .map_err(|_| Error::Unknown)?
                .with_no_client_auth();

            let https = hyper_rustls::HttpsConnectorBuilder::new()
                .with_tls_config(tls)
                .https_or_http()
                .enable_http1()
                .build();

            let client: Client<_, Full<Bytes>> = Client::builder(TokioExecutor::new()).build(https);
            let mut res = client.request(req).await.map_err(|_| Error::Unknown)?;
            if !res.status().is_success() {
                yield Err(Error::Unknown);
                return;
            }

            let mut buf = Vec::new();

            while let Some(frame) = res.frame().await {
                let frame = match frame {
                    Ok(f) if f.is_data() => f.into_data().unwrap(),
                    _ => continue
                };
                buf.extend_from_slice(&frame);

                // Each SSE event ends with \n\n
                while let Some(pos) = buf.windows(2).position(|w| w == b"\n\n") {
                    let event = buf.drain(..pos + 2).collect::<Vec<_>>();
                    let s = String::from_utf8_lossy(&event);
                    for line in s.lines() {
                        let line = line.trim();
                        if let Some(data) = line.strip_prefix("data: ") {
                            if data == "[DONE]" {
                                return;
                            }
                            let v = serde_json::from_str::<StreamChunk>(data);
                            if let Ok(chunk) = v {
                                // Adapt to StreamResponseChunk structure
                                let text = chunk.text
                                    .or_else(|| chunk.delta.as_ref().map(|d| d.text.clone()))
                                    .or_else(|| chunk.content_block.as_ref().map(|cb| cb.text.clone()));
                                if let Some(text) = text {
                                    let choice = CompletionChoice {
                                        message: None,
                                        text: Some(text),
                                        index: chunk.index.unwrap_or(0) as u16,
                                        delta: None,
                                        logprobs: None,
                                        finish_reason: None,
                                    };

                                    let response_chunk = StreamResponseChunk {
                                        id: format!("cmpl-{}", uuid::Uuid::new_v4()), // Generate a unique ID
                                        object: "chat.completion.chunk".to_string(),
                                        created: std::time::SystemTime::now()
                                            .duration_since(std::time::UNIX_EPOCH)
                                            .unwrap_or_default()
                                            .as_secs(),
                                        model: request.model.clone(),
                                        system_fingerprint: None,
                                        choices: vec![choice],
                                    };

                                    yield Ok(response_chunk);
                                }
                            }
                        }
                    }
                }
            }
        };

        Box::pin(s)
    }
}
